File: packages/core/src/interpret.ts
import type { ChoiceResponse, NoulResponse, SystemOneResult, Questions } from "@typesafe-ai/sdk";
import { COVERED, NOVEL, PICK, verifyKey } from "./questions.ts";
import type { Candidate, Decision, ResolveOptions } from "./types.ts";

function isChoice(a: unknown): a is ChoiceResponse {
  return (a as ChoiceResponse)?.type === "choice";
}
function isNoul(a: unknown): a is NoulResponse {
  return (a as NoulResponse)?.type === "noul";
}

/**
 * Turn Jev's fanned-out answers into apply/escalate. Policy lives here, in
 * code — the model supplies judgments, we decide what they permit.
 *
 * Escalate when: coverage fails (answer isn't in the set), the model itself
 * picked needs-novel-merge, the winner fails dual-intent verification, or
 * choice confidence is under threshold.
 */
export function interpret(
  result: SystemOneResult<Questions>,
  candidates: Candidate[],
  opts: ResolveOptions = {},
): Decision {
  const minConfidence = opts.minConfidence ?? 0.5;
  const minCoverage = opts.minCoverage ?? 0.5;
  const minVerify = opts.minVerify ?? 0.5;

  const answers = result.answers;
  const pick = answers[PICK];
  const covered = answers[COVERED];

  const verify: Record<string, number> = {};
  for (const c of candidates) {
    const a = answers[verifyKey(c.kind)];
    if (isNoul(a)) verify[c.kind] = a.noul;
  }

  const detail: Decision["detail"] = {
    verify,
    coverage: isNoul(covered) ? covered.noul : undefined,
  };
  if (isChoice(pick)) {
    detail.picked = pick.choice;
    detail.confidence = pick.confidence;
    detail.probabilities = { ...pick.probabilities };
  }

  const escalate = (reason: Decision["reason"]): Decision => ({ action: "escalate", reason, detail });

  if (isNoul(covered) && covered.noul < minCoverage) {
    return escalate("not-in-candidates");
  }
  if (!isChoice(pick)) return escalate("ask-failed");
  if (pick.choice === NOVEL) return escalate("novel-merge-needed");

  const winner = candidates.find((c) => c.kind === pick.choice);
  if (!winner) return escalate("ask-failed");

  if (verify[winner.kind] !== undefined && verify[winner.kind] < minVerify) {
    return escalate("verification-failed");
  }
  if (pick.confidence < minConfidence) return escalate("low-confidence");

  return { action: "apply", candidate: winner, detail };
}

File: packages/core-go/interpret.go
package core

import "sort"

// STRONG_FLOOR: below this a signal is a "confident no", not a hedge.
const strongFloor = 0.3

// Interpret turns Jev's fanned-out answers into apply/escalate. Policy
// lives here, in code — the model supplies judgments, we decide what they
// permit.
//
// Coverage, winner verification and pick confidence all hedge around ~0.5
// on genuinely ambiguous conflicts, so each alone is a noisy veto.
// Escalate only when signals concur: a strong single failure (< 0.3) or
// two or more weak failures (< their thresholds). A lone borderline signal
// is treated as hedging. A needs-novel-merge pick always escalates, as
// does a failed ask.
func Interpret(result *SystemOneResult, candidates []Candidate, opts *ResolveOptions) Decision {
	minConfidence := opts.minConfidence()
	minCoverage := opts.minCoverage()
	minVerify := opts.minVerify()
	minMargin := opts.minPickMargin()

	answers := result.Answers
	pick, pickOK := answers[PICK]
	covered, coveredOK := answers[COVERED]

	verify := map[string]float64{}
	for _, c := range candidates {
		if a, ok := answers[VerifyKey(c.Kind)]; ok && a.IsNoul() {
			verify[string(c.Kind)] = a.Noul
		}
	}

	detail := DecisionDetail{Verify: verify}
	if coveredOK && covered.IsNoul() {
		detail.Coverage = f64(covered.Noul)
	}
	pickIsChoice := pickOK && pick.IsChoice()
	if pickIsChoice {
		detail.Picked = pick.Choice
		detail.Confidence = f64(pick.Confidence)
		detail.Probabilities = map[string]float64{}
		for k, v := range pick.Probabilities {
			detail.Probabilities[k] = v
		}
		vals := make([]float64, 0, len(detail.Probabilities))
		for _, v := range detail.Probabilities {
			vals = append(vals, v)
		}
		sort.Sort(sort.Reverse(sort.Float64Slice(vals)))
		if len(vals) >= 2 {
			detail.PickMargin = f64(vals[0] - vals[1])
		}
	}

	escalate := func(reason EscalationReason) Decision {
		return Decision{Action: "escalate", Reason: reason, Detail: detail}
	}

	if !pickIsChoice {
		return escalate(ReasonAskFailed)
	}
	if pick.Choice == NOVEL {
		return escalate(ReasonNovelMerge)
	}

	var winner *Candidate
	for i := range candidates {
		if string(candidates[i].Kind) == pick.Choice {
			winner = &candidates[i]
			break
		}
	}
	if winner == nil {
		return escalate(ReasonAskFailed)
	}

	conf := pick.Confidence
	margin := 1.0
	if detail.PickMargin != nil {
		margin = *detail.PickMargin
	}

	type fail struct {
		reason EscalationReason
		value  float64
	}
	var fails []fail
	var ver *float64
	if v, ok := verify[string(winner.Kind)]; ok {
		ver = &v
	}
	if detail.Coverage != nil && *detail.Coverage < minCoverage {
		fails = append(fails, fail{ReasonNotInCandidates, *detail.Coverage / minCoverage})
	}
	if ver != nil && *ver < minVerify {
		fails = append(fails, fail{ReasonVerification, *ver / minVerify})
	}
	if conf < minConfidence {
		fails = append(fails, fail{ReasonLowConfidence, conf / minConfidence})
	}
	if margin < minMargin {
		fails = append(fails, fail{ReasonLowConfidence, margin / minMargin})
	}

	strong := (detail.Coverage != nil && *detail.Coverage < strongFloor) ||
		(ver != nil && *ver < strongFloor) ||
		conf < strongFloor

	if strong || len(fails) >= 2 {
		sort.SliceStable(fails, func(i, j int) bool { return fails[i].value < fails[j].value })
		return escalate(fails[0].reason)
	}

	return Decision{Action: "apply", Candidate: winner, Detail: detail}
}
